api/billingd/service/service.go

```diff
@@ -755,6 +771,59 @@ func (s *Service) DeleteCustomer(ctx context.Context, req *pb.DeleteCustomerRequ
 	return &pb.DeleteCustomerResponse{}, nil
 }
 
+func (s *Service) GetCustomerUsage(
+	ctx context.Context,
+	req *pb.GetCustomerUsageRequest,
+) (*pb.GetCustomerUsageResponse, error) {
+	doc, err := s.getCustomer(ctx, "_id", req.Key)
+	if err != nil {
+		return nil, err
+	}
+	usage := make(map[string]*pb.Usage)
+	for k, u := range doc.DailyUsage {
+		if product, ok := s.products[k]; ok {
+			// Get reported usage over the current invoice period
+			free, err := s.getPeriodUsageItem(u.FreeItemID)
+			if err != nil {
+				return nil, err
+			}
+			paid, err := s.getPeriodUsageItem(u.PaidItemID)
+			if err != nil {
+				return nil, err
+			}
+			total := (free.TotalUsage + paid.TotalUsage) * product.UnitSize
+			// Add current day unreported usage
+			if product.FreeQuotaInterval == FreeQuotaDaily &&
+				product.PriceType == PriceTypeIncremental {
+				// Add current day unreported usage
+				total += u.Total
+			}
+			usage[k] = getUsage(product, total, doc.InvoicePeriod)
+		}
+	}
+	return &pb.GetCustomerUsageResponse{
+		Usage: usage,
+	}, nil
+}
+
+func (s *Service) getPeriodUsageItem(id string) (sum *stripe.UsageRecordSummary, err error) {
+	params := &stripe.UsageRecordSummaryListParams{
+		SubscriptionItem: stripe.String(id),
+	}
+	params.Filters.AddFilter("limit", "", "1")
+	i := s.stripe.UsageRecordSummaries.List(params)
+	if i.Err() != nil {
+		return nil, i.Err()
+	}
+	for i.Next() {
+		sum = i.UsageRecordSummary()
+	}
```

Comment:
We need an extra `.Err()` check after the `for` loop, the docs say:
```
// Err returns the error, if any,                                                                                                                                                                                                                                                       
// that caused the Iter to stop.                                                                                                                                                                                                                                                        
/ It must be inspected                                                                                                                                                                                                                                                                 
// after Next returns false.  
```

Reply:
Ah good catch. Hard to tell if we need to check `.Err()` before `.Next()`, but doesn't look like it.

Reply:
Yep, not clear.